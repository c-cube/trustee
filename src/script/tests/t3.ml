module K = Kernel

let () =
  Printexc.register_printer (function
    | Error.E e -> Some (Error.show e)
    | _ -> None)

let s =
  {|

// naive fib
fn fib(n) {
  if n <= 2 {
    1
  } else {
    fib(n - 1) + fib(n - 2)
  }
}

// TODO: arrays + map with a closure

"fib 5";
fib(5);
fib(6);

"fib 12";
fib(12);

"fib 24";
fib(24);

|}

let ast =
  match parse_top_str ~filename:"t3" s with
  | Ok l -> l
  | Error e ->
    Format.printf "error: %a@." Error.pp e;
    []
;;

Format.printf "ast: %a@." Ast.pp_top ast

let ctx = K.Ctx.create ()

(* compile *)

let _env, stanzas = Compile.compile_l Compile.Env.empty ast;;

Format.printf "@[<2>compiled stanzas:@ %a@]@."
  (Fmt.Dump.list VM.Stanza.debug)
  stanzas

(* evaluate *)

let debug_hook vm i = ()
  (*
  ()
  Fmt.eprintf "@[<2>exec `%a`@ in %a@]@." VM.Instr.pp i VM.dump vm
        *)
  [@@ocaml.warning "-27"]
;;

try List.iter (VM.eval_stanza_pp ~debug_hook ctx) stanzas with _ -> ()

let rec fib n =
  if n <= 2 then
    1
  else
    fib (n - 1) + fib (n - 2)

let () =
  Fmt.printf "expected fib(5) = %d@." (fib 5);
  Fmt.printf "expected fib(12) = %d@." (fib 12);
  Fmt.printf "expected fib(24) = %d@." (fib 24);
  ()
