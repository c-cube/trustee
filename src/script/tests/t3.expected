ast: [(S_fn ("fib", ["n"],
         { bl_items =
           [(Bl_eval
               E_if {
                 test = (E_binop (Leq, (E_var "n"), (E_const (C_int 2))));
                 then_ = { bl_items = [(Bl_eval (E_const (C_int 1)))] };
                 elseif = [];
                 else_ =
                 (Some { bl_items =
                         [(Bl_eval
                             (E_binop (Plus,
                                (E_app ("fib",
                                   [(E_binop (Minus, (E_var "n"),
                                       (E_const (C_int 1))))
                                     ]
                                   )),
                                (E_app ("fib",
                                   [(E_binop (Minus, (E_var "n"),
                                       (E_const (C_int 2))))
                                     ]
                                   ))
                                )))
                           ]
                         })})
             ]
           }
         ));
       (S_eval (E_const (C_string "fib 5")));
       (S_eval (E_app ("fib", [(E_const (C_int 5))])));
       (S_eval (E_app ("fib", [(E_const (C_int 6))])));
       (S_eval (E_const (C_string "fib 12")));
       (S_eval (E_app ("fib", [(E_const (C_int 12))])));
       (S_eval (E_const (C_string "fib 24")));
       (S_eval (E_app ("fib", [(E_const (C_int 24))])))]
compiled stanzas:
  [(def `fib`
    chunk[1->1|1 local] {
      :instrs
       0        (rload 0) ; deref n
       1        (int 2)
       2        leq
       3        (jifn 6)
       4        (int 1) ; if
       5        (jmp 17)
       6        (rload 0) ; deref n
       7        (int 1)
       8        sub
       9        curch
       10       (call 1)
       11       (rload 0) ; deref n
       12       (int 2)
       13       sub
       14       curch
       15       (call 1)
       16       add
       17       ret
    });
   (eval
    (thunk
     :st (lazy
          chunk[0->1|0 local] {
            :instrs
             0        (lload 0) ; "fib 5"
             1        ret
            :locals
             0        "fib 5"
          })));
   (eval
    (thunk
     :st (lazy
          chunk[0->1|0 local] {
            :instrs
             0        (int 5)
             1        (lload 0) ; <chunk>
             2        (call 1)
             3        ret
            :locals
             0        chunk[1->1|1 local] {
                        :instrs
                         0        (rload 0) ; deref n
                         1        (int 2)
                         2        leq
                         3        (jifn 6)
                         4        (int 1) ; if
                         5        (jmp 17)
                         6        (rload 0) ; deref n
                         7        (int 1)
                         8        sub
                         9        curch
                         10       (call 1)
                         11       (rload 0) ; deref n
                         12       (int 2)
                         13       sub
                         14       curch
                         15       (call 1)
                         16       add
                         17       ret
                      }
          })));
   (eval
    (thunk
     :st (lazy
          chunk[0->1|0 local] {
            :instrs
             0        (int 6)
             1        (lload 0) ; <chunk>
             2        (call 1)
             3        ret
            :locals
             0        chunk[1->1|1 local] {
                        :instrs
                         0        (rload 0) ; deref n
                         1        (int 2)
                         2        leq
                         3        (jifn 6)
                         4        (int 1) ; if
                         5        (jmp 17)
                         6        (rload 0) ; deref n
                         7        (int 1)
                         8        sub
                         9        curch
                         10       (call 1)
                         11       (rload 0) ; deref n
                         12       (int 2)
                         13       sub
                         14       curch
                         15       (call 1)
                         16       add
                         17       ret
                      }
          })));
   (eval
    (thunk
     :st (lazy
          chunk[0->1|0 local] {
            :instrs
             0        (lload 0) ; "fib 12"
             1        ret
            :locals
             0        "fib 12"
          })));
   (eval
    (thunk
     :st (lazy
          chunk[0->1|0 local] {
            :instrs
             0        (int 12)
             1        (lload 0) ; <chunk>
             2        (call 1)
             3        ret
            :locals
             0        chunk[1->1|1 local] {
                        :instrs
                         0        (rload 0) ; deref n
                         1        (int 2)
                         2        leq
                         3        (jifn 6)
                         4        (int 1) ; if
                         5        (jmp 17)
                         6        (rload 0) ; deref n
                         7        (int 1)
                         8        sub
                         9        curch
                         10       (call 1)
                         11       (rload 0) ; deref n
                         12       (int 2)
                         13       sub
                         14       curch
                         15       (call 1)
                         16       add
                         17       ret
                      }
          })));
   (eval
    (thunk
     :st (lazy
          chunk[0->1|0 local] {
            :instrs
             0        (lload 0) ; "fib 24"
             1        ret
            :locals
             0        "fib 24"
          })));
   (eval
    (thunk
     :st (lazy
          chunk[0->1|0 local] {
            :instrs
             0        (int 24)
             1        (lload 0) ; <chunk>
             2        (call 1)
             3        ret
            :locals
             0        chunk[1->1|1 local] {
                        :instrs
                         0        (rload 0) ; deref n
                         1        (int 2)
                         2        leq
                         3        (jifn 6)
                         4        (int 1) ; if
                         5        (jmp 17)
                         6        (rload 0) ; deref n
                         7        (int 1)
                         8        sub
                         9        curch
                         10       (call 1)
                         11       (rload 0) ; deref n
                         12       (int 2)
                         13       sub
                         14       curch
                         15       (call 1)
                         16       add
                         17       ret
                      }
          })))]
(eval:
   "fib 5")
(eval:
   5)
(eval:
   8)
(eval:
   "fib 12")
(eval:
   144)
(eval:
   "fib 24")
(eval:
   46368)
expected fib(5) = 5
expected fib(12) = 144
expected fib(24) = 46368
